### PR TITLE
install: drop legacy telemetry install

### PR DESCRIPTION
**Pull request by howardjohn**
This has been around as deprecated usage for a few releases now, I think
its reasonable to finally drop all our builtin EnvoyFilter hacks.

This PR is fairly large.

* Primary change: remove EnvoyFilters for telemetry under manifests/
* Add additional validation to explicitly fail if they try to set (very advanced/niche) fields that required ^; these have _warned_ for a long time.
* Update operator "snapshots" tests, which required a few seemingly unrelated changes to tests.
* Refactor and fix tests in response to this
  * Fix the stackdriver tests so they properly detect on or off GKE; today, it gets confused if you are on a GCE VM running kind (or similar setups). This impacts our CI, which previously was using the real GCE metadata server; now it uses the fake. The fake had some issues which is now resolved.
  * Drop "access logging policy" test, which is only supported by EnvoyFilter and has no clear purpose today.
  * Consolidate "api" and "envoyfilter" suites into one single one (since envoyfilter is gone)
  * We have an "audit log" support in SD; this was done by EnvoyFilter. While we explore its long term future, put a temporary feature flag in istiod instead.
  * Remove metric _creation_ test, which is undocumented (on istio.io) and unsupported in Telemetry. Note the _customization_ is still supported + tested

### CODE REVIEW

**Comment by bochengchu** [comment]
Ah just saw this. But maybe we still need #47894 to patch older versions?

**Comment by howardjohn** [comment]
yep still need that one. Will let that go first, my PR needs some fixes
anyways

File: tests/integration/telemetry/stackdriver/testdata/server_access_log.json.tmpl [hunk 1]
```diff
@@ -26,7 +26,7 @@
         "source_version": "v1",
         "source_workload": "clt-{{ .ClusterName }}-v1",
         "protocol": "grpc",
-        "log_sampled": "true",
+        "log_sampled": "false",
```
Comment (howardjohn) [inline tests/integration/telemetry/stackdriver/testdata/server_access_log.json.tmpl:29]: this is from removal of "access log policy"

**Review by hzxuzhonghu — APPROVED** [review]

**Review by zirain — COMMENTED** [review]
I think this PR contains things more than I expected:
1. removal of stats EnvoyFilter
2. cleanup of metadata exchange
3. other chart refactor

that make this PR too huge to review, maybe we can do this one by one if it won't waste you too much time to split.

**Comment by howardjohn** [comment]
> that make this PR too huge to review, maybe we can do this one by one if it won't waste you too much time to split.

I was surprised when you initially said not to split it :slightly_smiling_face: . It grew a lot. Ill split it. This PR will eventually be the manifest changes, Ill open another PR with operator changes, and one with the test changes. I _think_ that should be sufficiently small but TBD

**Comment by zirain** [comment]
> > that make this PR too huge to review, maybe we can do this one by one if it won't waste you too much time to split.
> 
> I was surprised when you initially said not to split it :slightly_smiling_face: . It grew a lot. Ill split it. This PR will eventually be the manifest changes, Ill open another PR with operator changes, and one with the test changes. I _think_ that should be sufficiently small but TBD

I used to think everything in this PR just related 'drop stats envoyfilter', sorry for the misunderstand.

File: tests/integration/telemetry/stackdriver/testdata/server_access_log.json.tmpl (hunk 1 above)
Comment (kyessenov) [reply tests/integration/telemetry/stackdriver/testdata/server_access_log.json.tmpl]: Does ASM use this setting? We need some coverage - but not sure if this is tested here or elsewhere.

Comment (howardjohn) [reply tests/integration/telemetry/stackdriver/testdata/server_access_log.json.tmpl]: There's no docs, and it's impossible to use on MCP. I don't think there's any usage based on ~15min of digging

Comment (howardjohn) [reply tests/integration/telemetry/stackdriver/testdata/server_access_log.json.tmpl]: Seems like a half baked featured that was never completed from what I can tell

Comment (kyessenov) [reply tests/integration/telemetry/stackdriver/testdata/server_access_log.json.tmpl]: @lei-tang can you confirm? this has a high potential of a surprise bill if it's used and we miss it.

Comment (howardjohn) [reply tests/integration/telemetry/stackdriver/testdata/server_access_log.json.tmpl]: even if it is somehow configured, I am fairly sure it bills exactly the same. It's purely an opaque label .

Which doesn't make much sense, so either I am missing something or the feature was never really completed

Comment (kyessenov) [reply tests/integration/telemetry/stackdriver/testdata/server_access_log.json.tmpl]: The bill for access logs is substantial. Let's make sure that "Telemetry API" always configures "ERRORS_ONLY" on outbound which requires the sampler I think.

Comment (howardjohn) [reply tests/integration/telemetry/stackdriver/testdata/server_access_log.json.tmpl]: I'll double check what the mode is but just to clarify, this PR only changes behavior if someone is configured an undocumented and unsupported config to set access_logging_policy (not possible in MCP, either)

For anyone else there is no change at all. 

even if they did set it, though, it has nothing to do with billing as far as I can tell. It does *not* sample despite the name it only sets a bool flag that has no impact

Comment (lei-tang) [reply tests/integration/telemetry/stackdriver/testdata/server_access_log.json.tmpl]: "Set an access log policy" with incluster control plane is documented at: https://cloud.google.com/service-mesh/docs/observability/accessing-logs#set_an_access_log_policy.

If this feature is removed, documents need to be updated.

Comment (howardjohn) [reply tests/integration/telemetry/stackdriver/testdata/server_access_log.json.tmpl]: Thanks I did not find that during my search (largely because it never actually says *how* to configure it, so I assume no one actually knows).

Even with this, the feature does not work and isn't usable with MCP, so I cannot imagine anyone is using it. I would still encourage us to move forward with this PR as-is and I will take a follow-up to update the docs.

If there is a disagreement I can add it back, but we are maintaining a ton of code that does nothing

Comment (lei-tang) [reply tests/integration/telemetry/stackdriver/testdata/server_access_log.json.tmpl]: If this feature is not used by customers, it sounds good to me to remove this feature. If it has usage, I think a feature deprecation procedure may need to be followed.
 
Xinnan (@richardwxn) is the expert on feature usage data. Xinnan, do you know the usage data for this feature?

Comment (kyessenov) [reply tests/integration/telemetry/stackdriver/testdata/server_access_log.json.tmpl]: > Even with this, the feature does not work and isn't usable with MCP, so I cannot imagine anyone is using it. I would still encourage us to move forward with this PR as-is and I will take a follow-up to update the docs.

I vaguely remember a user explicitly requested it and is well aware how this works. We have to check the usage first.

Comment (howardjohn) [reply tests/integration/telemetry/stackdriver/testdata/server_access_log.json.tmpl]: Thanks for pushing on this - I think I was wrong.

I went through the docs, and they were incorrect initially, and now provide no information on how to configure it. Here is some options, in my order of preference:

1. Remove everything, as this PR does. Users can manually apply the EnvoyFilter if they need this.
2. Remove everything, as this PR does, but add a pilot env var to set this.
3. Keep just access_log_policy EnvoyFilter

In all cases I will also go back and add the tests back for this feature.

Comment (kyessenov) [reply tests/integration/telemetry/stackdriver/testdata/server_access_log.json.tmpl]: Access log policy only works with stackdriver, so it can also be tied to stackdriver provider API. But I'm fine with any installation time method or leave EnvoyFilter for it as-is.

Comment (howardjohn) [reply tests/integration/telemetry/stackdriver/testdata/server_access_log.json.tmpl]: Done, I did option (1) and added the test back with this enabled.

Comment (lei-tang) [reply tests/integration/telemetry/stackdriver/testdata/server_access_log.json.tmpl]: As with Option 1, user manually applies the EnvoyFilter. I think a user guide is needed, including the content of the EnvoyFilter for access_log_policy.

Comment (howardjohn) [reply tests/integration/telemetry/stackdriver/testdata/server_access_log.json.tmpl]: The user guide today already says "go figure it out". It doesn't say _how_ to do it at all. So I can improve the doc, but to be clear its already ~broken.

File: manifests/charts/istio-control/istio-discovery/templates/telemetryv2_1.21.yaml [hunk 2]
```diff
@@ -1,486 +0,0 @@
...
-        listener:
-          filterChain:
-            filter:
-              name: "envoy.filters.network.http_connection_manager"
-              subFilter:
-                name: "envoy.filters.http.router"
-      patch:
-        operation: INSERT_BEFORE
-        value:
-          name: istio.stats
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply manifests/charts/istio-control/istio-discovery/templates/telemetryv2_1.21.yaml:36]: These EnvoyFilters were already unused in 99% of scenarios and instead set via https://github.com/istio/istio/blob/cd981f42ba33497af1e5ec7b596838295e7f5355/manifests/charts/istio-control/istio-discovery/templates/configmap.yaml#L24 since 1.18. (fwiw on our MCP product, this has been the case since launch)

There is no change to behavior for almost all users. For the very very few users using audit log or access log policy (as discussed in the other thread), there is a slightly different enablement mechanism; the docs will be updated for this.

File: operator/pkg/apis/istio/v1alpha1/validation/validation.go [hunk 3]
```diff
@@ -107,21 +107,14 @@ func checkDeprecatedSettings(iop *v1alpha1.IstioOperatorSpec) (util.Errors, []st
...
 		{"Values.gateways.istio-ingressgateway.meshExpansionPorts", "components.ingressGateways[name=istio-ingressgateway].k8s.service.ports", nil},
 		{"AddonComponents.istiocoredns.Enabled", "the in-proxy DNS capturing (ISTIO_META_DNS_CAPTURE)", false},
 		{"Values.istiocoredns.enabled", "the in-proxy DNS capturing (ISTIO_META_DNS_CAPTURE)", false},
 		// nolint: lll
 		{"Values.global.jwtPolicy", "Values.global.jwtPolicy=third-party-jwt. See https://istio.io/latest/docs/ops/best-practices/security/#configure-third-party-service-account-tokens for more information", "third-party-jwt"},
-		{
-			"Values.telemetry.v2.stackdriver.logging", "Values.telemetry.v2.stackdriver.outboundAccessLogging and Values.telemetry.v2.stackdriver.inboundAccessLogging",
-			false,
-		},
+
```
Comment (lei-tang) [inline operator/pkg/apis/istio/v1alpha1/validation/validation.go:117]: Nit: the blank line 117 is not needed.

File: operator/cmd/mesh/testdata/manifest-generate/output/pilot_default.golden.yaml [hunk 4]
```diff
@@ -1849,14 +1849,15 @@ data:
                   successThreshold: 1
                   timeoutSeconds: 1
                 securityContext:
-                  privileged: true
+                  privileged: false
```
Comment (lei-tang) [inline operator/cmd/mesh/testdata/manifest-generate/output/pilot_default.golden.yaml:1852]: Is this related to "drop legacy telemetry install"? 
Same question for other changes under operator/cmd/mesh/testdata/manifest-generate/output.

File: operator/pkg/apis/istio/v1alpha1/validation/validation.go [hunk 5]
```diff
@@ -135,6 +128,18 @@ func checkDeprecatedSettings(iop *v1alpha1.IstioOperatorSpec) (util.Errors, []st
 		{"AddonComponents.tracing.Enabled", "the samples/addons/ deployments", false},
 		{"AddonComponents.kiali.Enabled", "the samples/addons/ deployments", false},
 		{"AddonComponents.prometheus.Enabled", "the samples/addons/ deployments", false},
+		{
+			"Values.telemetry.v2.stackdriver.logging", "Values.telemetry.v2.stackdriver.outboundAccessLogging and Values.telemetry.v2.stackdriver.inboundAccessLogging",
```
Comment (lei-tang) [inline operator/pkg/apis/istio/v1alpha1/validation/validation.go:132]: In line 131 to 133,
The "new" field in the deprecatedSettings structure is set as "Values.telemetry.v2.stackdriver.outboundAccessLogging and Values.telemetry.v2.stackdriver.inboundAccessLogging", which seems to be incorrect because outboundAccessLogging and inboundAccessLogging are removed by this PR.

File: operator/cmd/mesh/testdata/manifest-generate/output/pilot_default.golden.yaml (hunk 4 above)
Comment (howardjohn) [reply operator/cmd/mesh/testdata/manifest-generate/output/pilot_default.golden.yaml:1852]: No -- this is from 'Update operator "snapshots" tests, which required a few seemingly unrelated changes to tests.'

Basically, the tests here are based on a _snapshot_ of the manifests/ folder. This snapshot needs to be updated in this PR due to the size of the changes, but is not routinely update. So its pulling in changes from the past few weeks (at one point there was none, but this PR has been open for a while)

File: operator/pkg/apis/istio/v1alpha1/validation/validation.go (hunk 3 above)
Comment (howardjohn) [reply operator/pkg/apis/istio/v1alpha1/validation/validation.go:117]: thanks, fixed.

File: operator/pkg/apis/istio/v1alpha1/validation/validation.go (hunk 5 above)
Comment (howardjohn) [reply operator/pkg/apis/istio/v1alpha1/validation/validation.go:132]: Thanks! I fixed this

**Review by lei-tang — APPROVED** [review]

**Review by GregHanson — COMMENTED** [review]
I still see some references to stackdriver.configOverride in `pkg/config/analysis/analyzers/testdata/sidecar-injector-configmap-absolute-override.yaml`, `pkg/config/analysis/analyzers/testdata/sidecar-injector-configmap-with-revision-canary.yaml` , `pkg/config/analysis/analyzers/testdata/common/sidecar-injector-configmap.yaml` and `pkg/config/analysis/analyzers/testdata/common/sidecar-injector-enabled-nsbydefault.yaml`

**Review by GregHanson — APPROVED** [review]

**Review by zirain — APPROVED** [review]
LGTM, need a release notes for GKE(stackdriver) users?

**Review by hanxiaop — APPROVED** [review]

**Comment by hanxiaop** [comment]
> I still see some references to stackdriver.configOverride in `pkg/config/analysis/analyzers/testdata/sidecar-injector-configmap-absolute-override.yaml`, `pkg/config/analysis/analyzers/testdata/sidecar-injector-configmap-with-revision-canary.yaml` , `pkg/config/analysis/analyzers/testdata/common/sidecar-injector-configmap.yaml` and `pkg/config/analysis/analyzers/testdata/common/sidecar-injector-enabled-nsbydefault.yaml`

The analyzers' test files should have no issue with this.

**Review by stevenctl — APPROVED** [review]
